### PR TITLE
Particles in AMR

### PR DESCRIPTION
**Pull request by woolt**
Solve particle scenarios using the Piernik Adaptive Mesh Refinement structure and attribute particles to the finest levels of grid pieces and their boundaries

### CODE REVIEW

**Comment by gawrysz** [comment]
test this please

**Comment by gawrysz** [comment]
test this please

**Comment by gawrysz** [comment]
test this please

**Comment by gawrysz** [comment]
test this please

**Comment by gawrysz** [comment]
test this please

File: src/particles/particle_utils.F90 [hunk 1]
```diff
@@ -77,6 +75,10 @@ subroutine is_part_in_cg(cg, pos, indomain, in, phy, out, fin)
          in  = .true.
          phy = .true.
          out = .true.
+         ijk = ijk_of_particle(pos, dom%edge(:,LO), cg%idl)
+         ijk = max(ijk, cg%l%off)
+         ijk = min(ijk, cg%l%off + cg%l%n_d)
+         fin = cg%leafmap(ijk(xdim), ijk(ydim), ijk(zdim))
```
Comment (gawrysz) [inline src/particles/particle_utils.F90:79]: This can lead to reads way outside leafmap bounds when particles start to travel outside the domain

**Comment by gawrysz** [comment]
Test this please

**Comment by gawrysz** [comment]
test this please

File: src/particles/particle_utils.F90 (hunk 1 above)
Comment (woolt) [reply src/particles/particle_utils.F90:79]: fixed by [e32f032](https://github.com/piernik-dev/piernik/pull/532/commits/e32f0328c9844af34e4dc4b2ce3d804c7e829b23)

**Comment by gawrysz** [comment]
test this please

**Pull request merged**